packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -0,0 +1,460 @@
+      // Changed module directory
+      if ((children.size === 0 || hasInnerNodeModules) && wasModuleDirChanged(nodePath, stats, mtime, DEL_BLACKLIST, hasInnerNodeModules)) {
+        if (!hasInnerNodeModules) {
+          // Module has no inner node_modules, remove module dir
+          xfs.removeSync(nodePath);
+          // Persist module dir
+          persistDirEntry(parentPath, dir, nodeModulesFS, mtime, children);
+        } else {
+          // Module has inner node_modules, remove everything except inner node_modules
+          const dstEntryNames = xfs.readdirSync(nodePath);
+          for (const entry of dstEntryNames) {
+            const entryPath = ppath.join(nodePath, entry);
+            if (entry !== NODE_MODULES && !DEL_BLACKLIST.test(entryPath)) {
+              xfs.removeSync(entryPath);
+            }
+          }
+          // Persist everything, except inner node_modules
+          const srcEntries = nodeModulesFS.readdirSync(nodePath, {withFileTypes: true});
+          for (const entry of srcEntries) {
+            if (entry.name !== NODE_MODULES) {
+              persistDirEntry(nodePath, entry.name, nodeModulesFS, mtime, children, entry);
+            }
+          }
+          // Persist inner node_modules
+          const node = children.get(NODE_MODULES)!;
+          persistNode(nodePath, NODE_MODULES, node.mtime, node.children);
+        }
+        // Changed container directory
+      } else if (+stats.mtime !== +mtime) {
+        const srcEntryNames = new Set(children.keys());
+        const dstEntryNames = new Set(stats ? xfs.readdirSync(nodePath) : []);
+
+        for (const [entry, node] of children.entries()) {
+          if (!dstEntryNames.has(entry)) {
+            // Add new directories
+            persistDirEntry(nodePath, entry, nodeModulesFS, node.mtime, node.children);
+          } else {
+            // Check directories with the same name for changes
+            persistNode(nodePath, entry, node.mtime, node.children);
+          }
+        }
+
+        for (const entry of dstEntryNames) {
+          const entryDir = ppath.join(nodePath, entry);
+          if (!srcEntryNames.has(entry)) {
+            // Remove old directories
+            xfs.removeSync(entryDir);
+          }
+        }
+        xfs.utimesSync(nodePath, mtime, mtime);
+      } else {
+        // Unchanged container dir
+        for (const [entry, node] of children.entries()) {
+          persistNode(nodePath, entry, node.mtime, node.children);
+        }
+      }
+    }
+  };
+
+  let nmStats;
+
+  const rootNmDirPath = ppath.join(rootPath, NODE_MODULES);
+  try {
+    nmStats = nodeModulesFS.lstatSync(rootNmDirPath);
+  } catch (e) {
+    if (e.code !== 'ENOENT') {
+      throw e;
+    }
+  }
+
+  if (!nmStats) {
+    xfs.removeSync(rootNmDirPath);
+  } else {
+    const node = tsTree.get(NODE_MODULES)!;
+    persistNode(rootPath, NODE_MODULES, node.mtime, node.children);
+  }
+
+  return newTsTree;
```

Comment:
How does it work when workspaces need to have their own n_m folders? I would have expected this part of the code to call `persistNode` over each workspace root 🤔